org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/editor/contentassist/IdeContentProposalAcceptor.java

```diff
@@ -24,11 +27,25 @@
 public class IdeContentProposalAcceptor
 		implements IIdeContentProposalAcceptor, Comparator<Pair<Integer, ContentAssistEntry>> {
 
+	@Inject
+	private OperationCanceledManager operationCanceledManager;
+
+	protected CancelIndicator cancelIndicator;
+
+
 	protected final Set<Pair<Integer, ContentAssistEntry>> entries = new TreeSet<Pair<Integer, ContentAssistEntry>>(
 			this);
 
+	/**
+	 * @since 2.28
+	 */
+	public void setCancelIndicator(CancelIndicator cancelIndicator) {
+		this.cancelIndicator = cancelIndicator;
+	}
+
 	@Override
 	public void accept(ContentAssistEntry entry, int priority) {
+		operationCanceledManager.checkCanceled(cancelIndicator);
 		if (entry != null) {
 			if (entry.getProposal() == null)
 				throw new IllegalArgumentException("Proposal must not be null.");
```

Comment:
Cannot annotate an unchanged line:
I'm not sure if `canAcceptMoreProposals` is really being used, but maybe ask the cancelIndicator there if it really can accept more?
What happens if the request is cancelled - should it still answer with a partial result or should it cancel the entire future?

Reply:
looks like its up to the server what to do with cancellations
https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#cancelRequest

Reply:
@szarnekow , yes, `canAcceptMoreProposals` is used. I was unsure if it would be better to check for cancellation in `canAcceptMoreProposals` or in the `accept` method. I lean towards the current choice, but the other one would work as well and I do not mind changing it.

The current behaviour is to stop adding more proposals but the ones already added will be returned back.

That is what is done in
```
			try {
				createProposals(document.getContents(), position, caretOffset, resource, acceptor);
			} catch (Throwable t) {
				if (!operationCanceledManager.isOperationCanceledException(t)) {
					throw t;
				}
			}
			int idx = 0;
			for (ContentAssistEntry it : acceptor.getEntries()) {
				CompletionItem item = toCompletionItem(it, caretOffset, caretPosition, document);
				item.setSortText(Strings.padStart(Integer.toString(idx), 5, '0'));
				result.getItems().add(item);
				idx++;
			}
			return result;

```
If the `Throwable `is `operationCanceledManager.isOperationCanceledException`, then it is not throw, and `toCompletionItems` is called for the ones available.

Reply:
I'd say we preferably send the computed proposals and do this via `canAcceptMore`.

Reply:
Ah sorry, I misread. All good.